Node compatibility

kaicataldo opened the pull request.

kaicataldo commented on index.js:
```diff
@@ -4,7 +4,6 @@ import Engine from './src/Engine';
 
 var Focss = Class.extend({
   init: function(root, extensions, scoped) {
-    scoped = scoped === undefined ? true : !!scoped;
```
This check is now done in the Style constructor function

kaicataldo commented on src/Engine.js:
```diff
@@ -280,14 +261,14 @@ Engine = Class.extend({
     this.rules.push(rule);
 
-    if (this._state) {
+    if (this.style.hasStyleSheet && this._state) {
```
We may want to consider *not* auto processing on insert in the future, but this guard allows us to not a make this a breaking change.

mikesherov replied:
need to auto process on insert
although I guess that's not a real use case?

kaicataldo commented on test/behavior/toString.js:
```diff
@@ -0,0 +1,48 @@
+import Focss from '../..';
```
We had discussed not taking time to write tests for a Node environment in this PR, but I found that the quickest way to figure out what was failing in Node was just to run it. Since I had the tests anyway, figured I should include them. These three tests hit the three forks of logic we currently have for inserting Focss descriptors.

mikesherov replied:
ok, cool :-)

kaicataldo commented on util/css.js:
```diff
@@ -20,27 +20,11 @@ var isUnitlessNumber = {
     return (base || document).querySelectorAll(selector);
   },
 
-  matches: function(element, selector) {
```
This isn't used anywhere

kaicataldo commented on util/css.js:
```diff
@@ -96,6 +80,7 @@ export default {
     result += '}';
-    return result;
+
+    return result.trim();
```
Hadn't tested for multiple rules and noticed there are spaces between them. This only trims the whitespace around the entire rule (selector + body)

kaicataldo commented on util/detectScoped.js:
```diff
@@ -1,59 +1,61 @@
-  // IE doesn't allow inserting of '' as a styleRule
-  if (testSheet.addRule) {
-    testSheet.addRule('c', 'blink');
+export default {
```
Changed from an IIFE so that we can conditionally detect scoped based on if the styles are being created in the browser or in a Node env

mikesherov replied:
Where is this `check` being called now?

mikesherov replied:
nvm, got it. 👍

mikesherov commented on src/Style.js:
```diff
@@ -0,0 +1,67 @@
+}
+genId.i = 1;
+
+export default class Style {
```
Null object pattern? Not in love.

mikesherov commented on test/unit/index.js:
```diff
@@ -50,6 +50,8 @@ describe('Focss', function() {
 
   describe('#toString()', function() {
     it('returns a string of processed styles', function() {
+      fox = new Focss();
```
in the main afterEach, put `fox = null;`, it would've vcaught this bug

mikesherov commented on src/Engine.js:
```diff
@@ -280,14 +261,14 @@ Engine = Class.extend({
     }
     else {
-      this.sheet.insertRule(rule.getSelector(this._prefix) + '{}', i);
+      this.style.insertRule(rule.getSelector(this.style.prefix) + '{}', i);
```
I though we were nuking concept of prefix and scoped?

kaicataldo replied:
Oh, okay - happy to do that, but wasn't aware. Sounds good to me :+1:

mikesherov commented on src/Engine.js:
```diff
@@ -127,17 +106,19 @@ Engine = Class.extend({
-    }
+    var selector = rule.getSelector(this.style.prefix);
+
+    if (this.style.hasStyleSheet) {
```
this is a weak inference. Better to say `isBrowser`

mikesherov replied:
or even, `hasDom`

mikesherov replied:
and shouldn't this be a gaurd clause at the top instead?

mikesherov replied:
like it seems cleaner to put `typeof navigator` check at the top of this function.

mikesherov replied:
actually... at the top of `process`

kaicataldo replied:
Ah, you're right - I was trying to guard against `_process` being called by `insert` when Focss doesn't have access to the DOM but there's another guard in place for that so this is totally unnecessary. Makes sense to do the check in `process` anyway, though.

mikesherov commented on src/Style.js:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  insertRule(rule, index) {
+    if (!this.sheet) { return; }
```
why is this needed?

kaicataldo replied:
The idea for these was for Engine to not have to care about whether the style sheet existed or not - just returns a No op in the case that there is no DOM

mikesherov commented on src/Style.js:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  deleteRule(index) {
+    if (!this.sheet) { return; }
```
why is this needed?

mikesherov commented on src/Style.js:
```diff
@@ -0,0 +1,67 @@
+    return this.sheet[this.compat.rules];
+  }
+
+  get rulesLength() {
```
not a fan of all these getters.

hzoo commented on src/Engine.js:
```diff
@@ -29,40 +24,24 @@ var foreachSelectorRegex = /%forEach\(([^,]+),(.+)\)$/i;
 var toggleSelectorPsuedoRegex = /:(hover|active)/;
 var toggleSelectorClassRegex = /\.(__[^ :]+)/;
-var otherPrefixRegex = getVendorPrefixRegex();
+var otherPrefixRegex = typeof navigator === 'undefined' ? new RegExp() : getVendorPrefixRegex();
```
Do we want a comment here to explain that the check is for node compat?

kaicataldo replied:
Using `hasDom` now, so I think it's self-documenting

kaicataldo commented:
Okay, so rather than the null object pattern, I could also add a check into Engine like: 

```
function hasDom() {
  return typeof window !== 'undefined';
}
```

Then I could use that for all the guards/working around DOM-reliant behavior, including bailing on `process` immediately if Focss doesn't have access to the DOM.

Two questions:
 * Thoughts on keeping the Style class? Or do you think that should all be inlined into the Engine? I like the separation (easier for me to reason about).
 * The other idea @Attamusc and I discussed was not auto-processing on insert when there is state, but that's a breaking change and I felt like that should be a separate issue if we decide to do it.

mikesherov commented on src/Engine.js:
```diff
@@ -147,7 +128,7 @@ Engine = Class.extend({
     // insert them into a stylesheet will cause an exception to be thrown
     // @see: http://stackoverflow.com/questions/23050001/insert-multiple-css-rules-into-a-stylesheet
-    if (otherPrefixRegex.test(selector)) {
+    if (this.style.hasStyleSheet && otherPrefixRegex.test(selector)) {
```
this file engine should know if it's in a browser or not. it shouldn't just infer from stylesheet

mikesherov replied:
or at least defer to `this.style.getPrefixRegex`

kaicataldo commented on test/unit/index.js:
```diff
@@ -1,24 +1,39 @@
+    let fox;
+
+    expect(() => {
       fox = new Focss();
```
This instance needs to be managed independently so that we ensure that the stylesheet that is created is also cleaned up from the DOM - the current implementation of this test leaks state into later tests

kaicataldo commented on test/unit/src/Engine.js:
```diff
@@ -13,9 +13,14 @@ describe('Engine', function() {
     }).not.toThrow();
+    expect(engine).toBeDefined();
+
+    engine.destroy();
```
Same as https://github.com/behance/focss/pull/37/files#r72528747

kaicataldo commented on src/Engine.js:
```diff
@@ -299,33 +292,6 @@ Engine = Class.extend({
-  }
 });
 
-// ES6 future-proofing
```
We can get rid of this now that we have ES6 with Babel, right?

mikesherov replied:
No. We need this code so that Engine is iterable. What *could* be removed is just the `typeof Symbol` check. Not worth removing IMO

kaicataldo replied:
Got it

kaicataldo commented:
Updated - Style is a simple wrapper around the stylesheet now and Engine handles all the DOM-checking/management of the stylesheet. Also removed scoped styles/prefixes, as per comments.

mikesherov commented on src/Style.js:
```diff
@@ -0,0 +1,31 @@
+import detectScoped from '../util/detectScoped';
+
+export default class Style {
+  constructor(root = document.body) {
```
Why is there a default if root is always being passed in?

kaicataldo replied:
This matches the current behavior - I removed it initially because I was using whether or not a `root` parameter was passed in to determine if it was using the DOM or not. Now that we have the much more resilient `hasDom` check happening in `Engine`, we can have a default. That being said, I don't think we actually use it this way anywhere. Think I should get rid of it? Breaking chance, so figured I'd leave it in for now.

mikesherov replied:
but this constructor is always called with a value.

kaicataldo replied:
Right, but it could be `null`. I put the default value in Style because that's the only thing that cares about `root` - would it be clearer if the default value happened at the Focss constructor level?

kaicataldo replied:
Ah, right, sorry, remembering why I put that there - it's because we can't reference `document` unless we're in the browser, and this encapsulates all of it within `Style`

mikesherov replied:
you mean `undefined`. You can pass undefined and it's initialize to `document.body`.... got it.

kaicataldo replied:
Yep!

mikesherov commented on src/Rule.js:
```diff
@@ -37,12 +36,8 @@ var Rule = Class.extend({
-  getSelector: function(prefix) {
-    if (!prefix) { return this.computedSelector || this.selector; }
-
-    return this.specificity.map(function(part) {
```
Do we still need specificity?

kaicataldo replied:
Isn't this only necessary when there's the concept of a prefix (which I removed in this PR)?

mikesherov replied:
right, so I'm asking can we get rid of `this.specificity` and `util/specificity` completely?

kaicataldo replied:
Oh I see - you're right, it's not used anywhere else

mikesherov commented on util/detectScoped.js:
```diff
@@ -1,59 +1,61 @@
+
+    // return the object with the appropriate flags
+    return {
+      scopeSupported: scopeSupported,
```
do we still need this?

mikesherov commented on util/detectScoped.js:
```diff
@@ -1,59 +1,61 @@
+      scopeSupported: scopeSupported,
+      rules: DOMRules,
+      sheet: DOMStyle,
+      changeSelectorTextAllowed: changeSelectorTextAllowed
```
do we still need this?

mikesherov commented on util/detectScoped.js:
```diff
@@ -1,59 +1,61 @@
 // Scoped style support detection
```
comment now irrelevant, also please delete GH link

mikesherov commented:
Good work, almost there! Really like the cleanliness of the new `Style` object.

mikesherov commented:
👍

mikesherov commented:
Breaking change, so this would be 2.0

hzoo commented:
👍
